### PR TITLE
[Intel MKL] Adding support to public CI for AVX512 builds for various versions of gcc

### PR DESCRIPTION
**Pull request by claynerobison**
Manually specifying each instruction on the command line was hurting performance in instances where gcc suppressed support for older SIMD instructions when newer instructions were used (e.g. don't use SSE when AVX instructions are available). This commit makes use of gcc's built-in platform-based optimizations rather than rolling our own.

Also, the older versions of gcc didn't support newer platform code names, so this commit adds support for detecting the version of gcc in the environment and then adding platform support in the idiom that version of gcc comprehends.

### CODE REVIEW

**Comment by claynerobison** [comment]
Hi @penpornk. Do you have a few minutes to review this PR?

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py [hunk 1]
```diff
@@ -54,53 +31,173 @@
...
+    self.host_gcc_minor_version_ = gcc_minor_version
+    return True
+
+  # return a string with all the necessary bazel formatted flags for this
+  # platform in this gcc environment
+  def get_bazel_gcc_flags(self):
+    raise NotImplementedError(self)
+
+  # return true or false depending on whether the host gcc version
+  # is newer or older than the gcc version in which the new
```
Comment (penpornk) [inline tensorflow/tools/ci_build/linux/mkl/set-build-env.py]: This returns true when the host version is older (not newer). Please update the description.

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py [hunk 2]
```diff
@@ -54,53 +31,173 @@
...
+    else:
+      return self.BAZEL_PREFIX_ + self.ARCH_PREFIX_ + \
+        SANDYBRIDGE_ARCH_NEW + " "
+
+class HaswellPlatform(IntelPlatform):
+  def __init__(self):
+    IntelPlatform.__init__(self, 4, 8)
+
+  def get_bazel_gcc_flags(self):
+    HASWELL_ARCH_OLD = "core-avx2" # Only missing the POPCNT instruction
```
Comment (penpornk) [inline tensorflow/tools/ci_build/linux/mkl/set-build-env.py]: Nit:
```suggestion
    HASWELL_ARCH_OLD = "core-avx2"  # Only missing the POPCNT instruction
```

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py [hunk 3]
```diff
@@ -54,53 +31,173 @@
...
+
+class SkylakePlatform(IntelPlatform):
+  def __init__(self):
+    IntelPlatform.__init__(self, 4, 9)
+
+  def get_bazel_gcc_flags(self):
+    SKYLAKE_ARCH_OLD = "broadwell"  # Only missing the POPCNT instruction
+    SKYLAKE_ARCH_NEW = "skylake-avx512"
+    # the flags that broadwell is missing: pku, clflushopt, clwb, avx512vl,
+    # avx512bw, avx512dq. xsavec and xsaves are availalbe on 5.x
```
Comment (penpornk) [inline tensorflow/tools/ci_build/linux/mkl/set-build-env.py]:
```suggestion
    # avx512bw, avx512dq. xsavec and xsaves are available on 5.x
```

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py [hunk 4]
```diff
@@ -186,13 +269,24 @@ def parse_args(self):
...
-        raise ValueError("{} is not a valid file name".format(
+        print ("You can't write bazel config to \"{}\" "
+          "because it is a directory".format(
             self.args.bazelrc_file))
+        return False
+
+    # Validate gcc with the requested platform
+    gcc_major_version, gcc_minor_version = self.get_gcc_version()
+    if gcc_major_version == 0 or \
+       False == self.target_platform_.set_host_gcc_version(
```
Comment (penpornk) [inline tensorflow/tools/ci_build/linux/mkl/set-build-env.py]: Nit:
```suggestion
       not self.target_platform_.set_host_gcc_version(
```

**Review by penpornk — CHANGES_REQUESTED** [review]
I'm so sorry for the delay! I have minor comments.

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py (hunk 1 above)
Comment (claynerobison) [reply tensorflow/tools/ci_build/linux/mkl/set-build-env.py]: Thank you for keeping me honest and precise!

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py (hunk 3 above)
Comment (claynerobison) [reply tensorflow/tools/ci_build/linux/mkl/set-build-env.py]: Good catch!

File: tensorflow/tools/ci_build/linux/mkl/set-build-env.py (hunk 4 above)
Comment (claynerobison) [reply tensorflow/tools/ci_build/linux/mkl/set-build-env.py]: Fair enough. More elegant; less c++ like.

**Review by penpornk — APPROVED** [review]
Thank you very much for the quick fixes!

**Comment by gbaned** [comment]
@claynerobison Can you please check build failures? Thanks!

**Review by penpornk — CHANGES_REQUESTED** [review]
@claynerobison Could you please help fix [Ubuntu Sanity failures](https://source.cloud.google.com/results/invocations/553fc495-eecf-492d-81b5-3b148f92b8aa/log)?
(Search for "FAIL: Found 19 non-whitelisted pylint errors:". Running `pylint` should fix them.)

Other failures, e.g., `Ubuntu CPU`,  seem unrelated. I'll rerun the tests again after the `Ubuntu Sanity` fix.

**Review by penpornk — DISMISSED** [review]
Thank you for the fixes!

**Comment by gbaned** [comment]
@claynerobison Could you please help fix Ubuntu Sanity failures? Thanks!

**Review by penpornk — APPROVED** [review]
Thank you for your patience!

**Pull request merged**